### PR TITLE
Support on prem GitHub release notes / changelog 

### LINKED ISSUE #1088: Compare url & release notes/changelog are restricted to {github,gitlab}.com
Current logic of generating compare url and release notes/changelog are restricted to either github.com or gitlab.com
```
if (repoUrl.startsWith("https://github.com/") || repoUrl.startsWith("https://gitlab.com/"))
```

Is it possible to configure/extend this, so that git enterprise urls (such as git.enterprise.company.com) can also be included in this?

### PR DESCRIPTION
**Pull request by eugeniyk**
Currently we have `github.com` and other urls hardcoded in the code, which makes it unusable in environments with different on-prem domain names.

Fixes #1088

### CODE REVIEW

File: modules/core/src/test/scala/org/scalasteward/core/vcs/VCSPackageTest.scala [hunk 1]
```diff
@@ -114,9 +113,4 @@ class VCSPackageTest extends AnyFunSuite with Matchers {
           "https://bitbucket.org/foo/bar/compare/release-1.2.3..release-1.2.0#diff"
         )
   }
-
-  test("possibleChangelogUrls: homepage") {
-    possibleReleaseRelatedUrls(uri"https://scalacenter.github.io/scalafix/", update)
-      .map(_.url.renderString) shouldBe List()
-  }
```
Comment (eugeniyk) [inline modules/core/src/test/scala/org/scalasteward/core/vcs/VCSPackageTest.scala:121]: not sure we need this?

File: modules/core/src/test/scala/org/scalasteward/core/vcs/VCSPackageTest.scala [hunk 2]
```diff
@@ -25,38 +26,36 @@ class VCSPackageTest extends AnyFunSuite with Matchers {
...
     )
-    possibleCompareUrls(uri"https://bitbucket.org/foo/bar", update)
+    possibleCompareUrls(SupportedVCS.Bitbucket, uri"https://bitbucket.org/foo/bar", update)
       .map(_.url.renderString) shouldBe List(
       "https://bitbucket.org/foo/bar/compare/v1.2.3..v1.2.0#diff",
       "https://bitbucket.org/foo/bar/compare/1.2.3..1.2.0#diff",
       "https://bitbucket.org/foo/bar/compare/release-1.2.3..release-1.2.0#diff"
     )
-
-    possibleCompareUrls(uri"https://scalacenter.github.io/scalafix/", update) shouldBe List()
```
Comment (eugeniyk) [inline modules/core/src/test/scala/org/scalasteward/core/vcs/VCSPackageTest.scala:55]: not sure we need this?

**Comment by mzuehlke** [comment]
Your change would restrict linking to release notes to only the VCS the current installation is running on.
In the current implementation the scala steward instance running on Gitlab or even on-prem could link to release notes on e.g. `github.com`. After your change this would not be possible anymore.

**Comment by eugeniyk** [comment]
@mzuehlke I've got your point, I thought it's trying to compare within current github installation.
currently our github installation uses "github.domain.io", making it's impossible to link CHANGELOG for local dependencies

Seems #1088 describes exactly my problem

**Comment by eugeniyk** [comment]
@mzuehlke looks like we can solve it this way:
1. if `repoUri` starts with `Config.vcsApiHost` then we can use `Config.vcsType` and apply current PR changes
2. otherwise fallback to the current implementation

WDYT?

**Comment by mzuehlke** [comment]
I think your proposal looks like a perfect way to solve this issue. 👍

**Comment by eugeniyk** [comment]
image

what if codacy starts to complain about old test cases and literal constants in tests? Do you want to apply those rules there as well?

**Comment by eugeniyk** [comment]
@mzuehlke please review again

**Comment by eugeniyk** [comment]
@fthomas could you please review this PR? The [Codacy failure](https://github.com/scala-steward-org/scala-steward/pull/1563#issuecomment-671079864) sounds like false positive

File: modules/core/src/main/scala/org/scalasteward/core/vcs/package.scala [hunk 3]
```diff
@@ -73,51 +73,90 @@ package object vcs {
     possibleFilenames(baseNames)
   }
 
-  def possibleCompareUrls(repoUrl: Uri, update: Update): List[VersionDiff] = {
+  private[this] def extractRepoVCSType(
```
Comment (mzuehlke) [inline modules/core/src/main/scala/org/scalasteward/core/vcs/package.scala:76]: one more, related to the first comment:
`extractRepoVCSType` could live in `VCSExtraAlg` and you would only need to pass the `RepoVCSType`.

**Review by mzuehlke — APPROVED** [review]
A really good PR.
Let me know what you think about my comments.
But I'm fine with merging this PR like it is now.

File: modules/core/src/main/scala/org/scalasteward/core/vcs/package.scala (hunk 3 above)
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/vcs/package.scala:76]: then `possibleCompareUrls` and `possibleReleaseRelatedUrls` would need to be extracted to `VCSExtraAlg` as well (right now it's in package object), which make sense for me as well

I can try to make it as separated PR

File: modules/core/src/main/scala/org/scalasteward/core/vcs/VCSExtraAlg.scala [hunk 4]
```diff
@@ -19,12 +19,18 @@ package org.scalasteward.core.vcs
...
+import org.scalasteward.core.application.SupportedVCS
 import org.scalasteward.core.data.{ReleaseRelatedUrl, Update}
 import org.scalasteward.core.util.HttpExistenceClient
 import org.scalasteward.core.vcs
 
 trait VCSExtraAlg[F[_]] {
-  def getReleaseRelatedUrls(repoUrl: Uri, update: Update): F[List[ReleaseRelatedUrl]]
+  def getReleaseRelatedUrls(
+      vcsType: SupportedVCS,
+      vcsUri: Uri,
```
[the remark this answers is not in the document]
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/vcs/VCSExtraAlg.scala:30]: Yes, can do that

[the remark this answers is not in the document]
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/vcs/VCSExtraAlg.scala:30]: ok, changed

File: modules/core/src/main/scala/org/scalasteward/core/vcs/package.scala (hunk 3 above)
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/vcs/package.scala:76]: @mzuehlke have it as separated PR here: https://github.com/eugeniyk/scala-steward/pull/1
I think we can merge #1563 as is and continue with refactorings in PR above, wdyt?

**Comment by mzuehlke** [comment]
Thanks @eugeniyk for contributing to scala-steward 👍

**Comment by eugeniyk** [comment]
@mzuehlke glad to improve this awesome tool!

**Pull request merged**